File: brainfuck.py
import sys
import getch

cells = [0] * 10000

f = open(sys.argv[1], 'r')
code = f.read()

pointer = 0
code_i = 0
brackets = {}

matching_queue = []
for i, c in enumerate(code):
    if c == '[':
        matching_queue.append(i)
    elif c == ']':
        opening_bracket = matching_queue.pop()
        brackets[i] = opening_bracket
        brackets[opening_bracket] = i

while code_i < len(code):
    c = code[code_i]
    if c == '>':
        pointer += 1
    elif c == '<':
        pointer -= 1
    elif c == '+':
        cells[pointer] += 1
    elif c == '-':
        cells[pointer] -= 1
    elif c == '.':
        print(chr(cells[pointer]), end='')
    elif c == ',':
        cells[pointer] = ord(getch.getch())
    elif c == '[' and cells[pointer] == 0:
        code_i = brackets[code_i]
    elif c == ']' and cells[pointer] != 0:
        code_i = brackets[code_i]
    code_i += 1

File: brainfuck.nim
import std/enumerate
import std/tables
import std/terminal
import std/os

let code = readFile(paramStr(1))

var cells: array[30000, int]
var cells_i = 0
var code_i = 0

var loops = initTable[int, int]()
var loop_matches: seq[int]
for i, c in enumerate(code):
  if c == '[':
    loop_matches.add(i)
  elif c == ']':
    let loop_start_i = loop_matches.pop()
    loops[i] = loop_start_i
    loops[loop_start_i] = i

while code_i < len(code):
  var c = code[code_i]
  case c:
    of '>':
      cells_i += 1
    of '<':
      cells_i -= 1
    of '+':
      cells[cells_i] += 1
    of '-':
      cells[cells_i] -= 1
    of '.':
      stdout.write(chr(cells[cells_i]))
    of ',':
      cells[cells_i] = ord(getch())
    of '[':
      if cells[cells_i] == 0:
        code_i = loops[code_i]
    of ']':
      if cells[cells_i] != 0:
        code_i = loops[code_i]
    else: discard
  code_i += 1
